src/NuGet.Clients/NuGet.SolutionRestoreManager/InfoBarService.cs

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using NuGet.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Imaging;
+using Microsoft.VisualStudio;
+using System.Threading.Tasks;
+using System.Threading;
+using NuGet.VisualStudio.Telemetry;
+
+#nullable enable
+
+namespace NuGet.SolutionRestoreManager
+{
+    [Export(typeof(IInfoBarService))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class InfoBarService : IInfoBarService, IVsInfoBarUIEvents
+    {
+        IAsyncServiceProvider _asyncServiceProvider = AsyncServiceProvider.GlobalProvider;
+        private IVsInfoBarUIElement? _infoBarUIElement;
+        private bool _visible;
+        private bool _closed = false;
+        private bool _closeFromHide = false;
+        private uint? _eventCookie;
+
+        [Import]
+        private Lazy<IPMUIStarter>? PackageManagerUIStarter { get; set; }
+
+        public async Task Show(CancellationToken cancellationToken)
+        {
+            await CreateAndShowInfoBarAsync(cancellationToken);
+        }
+
+        public async Task Hide(CancellationToken cancellationToken)
+        {
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            if (!_visible)
+            {
+                return;
+            }
+
+            _closeFromHide = true;
+
+            try
+            {
+                _infoBarUIElement?.Close();
+                _infoBarUIElement = null;
```

Comment:
Would it be more appropriate to `Dispose` here? If another vulnerability is detected in a future restore, a new `_infoBarUIElement` is being created, so I don't think we need the reference after this point.

Reply:
Generally, if an old instance isn't being used any longer, I think it's good to set the field to `null`, so that it can be GCd.

However, VS servers are (almost) always singletons, so letting go of the reference and needing to query it from the service provider again is basically a waste of CPU cycles.

Reply:
It might waste CPU cycles, but is it actually likely to happen that often? The scenario would be, you have vulnerabilities, you fix them, then you re-introduce new vulnerabilities. I would imagine that happens very seldom, or if it does then it's with diminishing returns.

My concern was more about memory. Holding InfoBar elements/hosts in memory for the remainder of the session when the probability is high that it will won't be used again.

Reply:
Question about this, `_infoBarUIElement` doesn't have the `Dispose` method, do I need to create one? I looked at index to look for examples and no ones uses Dispose, they only use Unadvise

Reply:
No, I don't think you need to create a `Dispose`. I thought last week when I looked at it, that `Unadvise` implemented `Dispose`. If it doesn't, then nevermind. If it does, then should you be calling `Unadvise` here? That was the basis of my question.